test_integration/base_suite_test.go

```diff
@@ -0,0 +1,285 @@
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+)
+
+var globalTestExecutionId string
+
+func TestMain(m *testing.M) {
+	globalTestExecutionId = strings.ToLower(strings.Split(uuid.New().String(), "-")[0])
+	log.Println("====> Starting the whole test flow: ", globalTestExecutionId)
+
+	err := setTestGlobalEnvVariables(map[string]string{})
+	if err != nil {
+		log.Println("Error Setting up Test Env Variables: ", err.Error())
+		os.Exit(1)
+	}
+
+	ctx := context.Background()
+	err = compileBacalhau(ctx, "../main.go")
+	if err != nil {
+		log.Println("Error compiling the bacalhau binary: ", err.Error())
+		os.Exit(1)
+	}
+
+	// TODO: Maybe we do not need to created images, but just inject
+	// TODO: them with artifacts before container starts the starts (certs and binary and configs)
+	err = buildBaseImages(globalTestExecutionId)
+	if err != nil {
+		log.Println("Error building base images: ", err.Error())
+		os.Exit(1)
+	}
+
+	exitCode := m.Run()
+
+	err = deleteDockerTestImagesAndPrune(globalTestExecutionId)
+	if err != nil {
+		log.Println("Error cleaning up base images: ", err.Error())
+		os.Exit(1)
+	}
+
+	// TODO: Better cleaning
+	os.Remove("./assets/dockerfiles/bacalhau_bin")
+	//Exit with the same code as the test run
+	os.Exit(exitCode)
+}
+
+type BaseDockerComposeTestSuite struct {
+	suite.Suite
+	ComposeStack           interface{}
+	AdditionalSetupEnvVars map[string]string
+	Context                context.Context
+	Cancel                 context.CancelFunc
+	suiteRunIdentifier     string
+	globalRunIdentifier    string
+}
+
+func (s *BaseDockerComposeTestSuite) SetupSuite(dockerComposeFilePath string, adhocImagesNames map[string]string) {
+	s.T().Log("Setting up [Test Suite] in base suite......")
+
+	// Merge Docker images name
+	initialImagesNames := map[string]string{
+		"placeholder-registry-image":  fmt.Sprintf("bacalhau-test-registry-%s:%s", s.globalRunIdentifier, s.globalRunIdentifier),
+		"placeholder-requester-image": fmt.Sprintf("bacalhau-test-requester-%s:%s", s.globalRunIdentifier, s.globalRunIdentifier),
+		"placeholder-compute-image":   fmt.Sprintf("bacalhau-test-compute-%s:%s", s.globalRunIdentifier, s.globalRunIdentifier),
+		"placeholder-jumpbox-image":   fmt.Sprintf("bacalhau-test-jumpbox-%s:%s", s.globalRunIdentifier, s.globalRunIdentifier),
+	}
+
+	if adhocImagesNames != nil {
+		for key, value := range adhocImagesNames {
+			initialImagesNames[key] = value
+		}
+	}
+
+	// Render Docker Compose File
+	tmpDir, err := os.MkdirTemp("", "bacalhau-test-"+s.suiteRunIdentifier)
+	s.Require().NoErrorf(err, "Error creating tmp dir: %q", err)
+
+	renderedDockerComposeFilePath, err := s.renderDockerComposeFile(
+		dockerComposeFilePath,
+		tmpDir,
+		initialImagesNames,
+	)
+	s.Require().NoErrorf(err, "Error rendering Docker Compose file: %q", err)
+
+	defer func() {
+		if err := os.RemoveAll(tmpDir); err != nil {
+			fmt.Printf("Error removing temporary directory: %v\n", err)
+		}
+	}()
```

Comment:
should this be done before `renderDockerComposeFile`?

Reply:
`renderDockerComposeFile` needs the common suite tmp dir, thats why I am defering its deletion till the end.

Reply:
understood, my point here was: if (in the rare case) `renderDockerComposeFile` returns an error, the `tmpDir` won't be cleaned up.

Reply:
I got you, good catch.